### PR TITLE
Add c shim wrapper codegen.

### PR DESCRIPTION
**Pull request by etaf**
Add c shim wrapper codegen. 
Support generate c shim wrapper for OPs in torch_xpu_ops which extend the c shim wrapper of Pytorch in-tree XPU ops.

### CODE REVIEW

**Comment by etaf** [comment]
This PR is relied upon by AOT Inductor PR https://github.com/pytorch/pytorch/pull/139026.
And should be able to pass the CI after Pytorch PR https://github.com/pytorch/pytorch/pull/136742 landed.
The design is descripted in https://github.com/pytorch/pytorch/pull/136742

File: cmake/Codegen.cmake [hunk 1]
```diff
@@ -76,11 +76,15 @@ function(GEN_XPU file_yaml)
     --per-operator-headers
     --static-dispatch-backend
     --backend-whitelist=XPU
+    --xpu
+    --update-aoti-c-shim
+    --aoti-extend
+    --aoti-install-dir ${XPU_AOTI_INSTALL_DIR}
```
Comment (EikanWang) [inline cmake/Codegen.cmake:82]: Are the two parameters newly added?

Comment (etaf) [reply cmake/Codegen.cmake:82]: --aoti-extend is add in https://github.com/pytorch/pytorch/pull/136742

**Comment by EikanWang** [comment]
CI failure. Because it depends on PyTorch PR, right?

**Review by EikanWang — APPROVED** [review]
LGTM

File: cmake/Codegen.cmake [hunk 2]
```diff
@@ -76,11 +76,21 @@ function(GEN_XPU file_yaml)
     --per-operator-headers
     --static-dispatch-backend
     --backend-whitelist=XPU
+    # --xpu: generate in-tree RegisterXPU.cpp for in-tree OPs
```
Comment (fengyuan14) [inline cmake/Codegen.cmake:79]: We will have two `RegisterXPU.cpp` ?

File: cmake/Codegen.cmake [hunk 3]
```diff
@@ -46,10 +46,11 @@ endfunction(GEN_BACKEND)
 
 set(RegisterXPU_PATH ${BUILD_TORCH_XPU_ATEN_GENERATED}/RegisterXPU.cpp)
 set(XPUFallback_PATH ${TORCH_XPU_OPS_ROOT}/src/ATen/native/xpu/XPUFallback.template)
+set(XPU_AOTI_INSTALL_DIR ${TORCH_ROOT}/torch/csrc/inductor/aoti_torch/generated/extend)
```
Comment (fengyuan14) [inline cmake/Codegen.cmake:49]: Can we use PyTorch defined CMAKE var for the path?

File: cmake/Codegen.cmake (hunk 2 above)
Comment (etaf) [reply cmake/Codegen.cmake:79]: @fengyuan14  Yes, in stock pytorch, we've already supported structured kernel generation for GEMM ops. The PR https://github.com/pytorch/pytorch/pull/139025 has been landed.

File: cmake/Codegen.cmake [hunk 4]
```diff
@@ -96,13 +106,20 @@ endfunction(GEN_XPU)
...
+ ${XPU_AOTI_INSTALL_DIR}/c_shim_xpu.cpp
 )
 
 
-
+# The c_shim_xpu.cpp needs include files in ${CMAKE_BINARY_DIR}/xpu/ATen/ops/*.h)
+# The include path is auto generated as "#include <ATen/ops/*.h">
+# To follow the design of aoti codegen, here ${CMAKE_BINARY_DIR}/xpu is added to
+# $TORCH_XPU_OPS_INCLUDE_DIRS, so that "#include <ATen/ops/*.h>" works.
+list(APPEND TORCH_XPU_OPS_INCLUDE_DIRS ${CMAKE_BINARY_DIR}/xpu)
```
Comment (fengyuan14) [inline cmake/Codegen.cmake:120]: Will the `<ATen/ops/*.h>` be ambiguous? Like we might have `native/ATen/ops/*.h`, or `cpu/ATen/ops.*.h`.

Comment (etaf) [reply cmake/Codegen.cmake:120]: The `c_shim_xpu.h/c_shim_xpu.cpp` and it's required headers `ATen/ops/*.h` is generated at the same script, so it always can find the dependency. we just match install dir of the torchgen.gen here.

File: cmake/Codegen.cmake (hunk 3 above)
Comment (etaf) [reply cmake/Codegen.cmake:49]: @fengyuan14 there is no available CMAKE var for the path.

**Comment by etaf** [comment]
@fengyuan14 can I merge this PR?

**Review by fengyuan14 — APPROVED** [review]

**Pull request merged**